### PR TITLE
[kernel-spark] Implement SupportsPushDownLimit in Delta V2 connector

### PR DESCRIPTION
**Pull request by rapoth**
- [x] Spark
- [x] Kernel

## Description

Implements Spark's `SupportsPushDownLimit` interface in the Delta V2 (kernel-based) connector, addressing #6326.

When a query like `SELECT * FROM delta_table LIMIT N` is executed, Spark pushes the limit into the scan. The connector then requests file-level statistics via the new `Scan.getScanFiles(engine, true)` API (#6330), accumulates logical row counts (numRecords minus deletion vector cardinality) per file, and stops adding files once enough rows are accumulated. This significantly reduces I/O for large tables with small limits.

**Key design decisions:**
- `pushLimit()` always returns `true`. The limit is a hint, not a guarantee. `isPartiallyPushed()` is always `true` because file-level granularity means the scan may return more rows than requested.
- Files without stats are conservatively included but not counted toward the limit, ensuring correctness over performance.
- `LIMIT 0` short-circuits immediately (returns empty iterator).
- `equals()`/`hashCode()` include `pushedLimit` for correct Spark plan cache behavior.

Depends on #6330 (kernel API to expose `getScanFiles(Engine, boolean)`).

## How was this patch tested?

28 new tests added across 3 test classes:

- **`V2LimitPushdownTest`** (15 E2E tests): basic correctness, deletion vectors, heavy deletion vectors, empty table, limit 0, limit 1, limit exceeding table size, partition filter + limit, data filter + limit, column mapping, streaming (verifies limit is NOT pushed for streaming), ORDER BY interaction, V1/V2 comparison, and no-stats fallback.
- **`SparkScanTest`** (8 unit tests): `planScanFiles` with limit, without limit, limit 0 short-circuit, files without stats, DV cardinality subtraction, description format, equals/hashCode with different limits.
- **`SparkScanBuilderTest`** (5 unit tests): `pushLimit` return value, `isPartiallyPushed`, limit propagation to scan, no-limit default, limit 0 handling.

All tests pass with Java 17:
```
build/sbt sparkV2/test
```

## Does this PR introduce _any_ user-facing changes?

Yes. `SELECT ... LIMIT N` queries against Delta tables using the V2 connector (`dsv2.delta`) now benefit from limit pushdown, reading fewer files when possible. This is transparent to users and requires no configuration. The optimization is best-effort: if file statistics are unavailable, the full scan is performed as before.

### CODE REVIEW

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java [hunk 1]
```diff
@@ -407,31 +423,82 @@ public String getTablePath() {
...
-          final Row row = addFileRowIter.next();
-          final AddFile addFile = new AddFile(row.getStruct(0));
-
-          final PartitionedFile partitionedFile =
-              PartitionUtils.buildPartitionedFile(addFile, partitionSchema, tablePath, zoneId);
+    // Short-circuit: LIMIT 0 means no files needed
+    if (pushedLimit.isPresent() && pushedLimit.getAsInt() == 0) {
+      estimatedSizeInBytes = 0;
+      return;
+    }
```
Comment (raveeram-db) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java:435]: please add soem logging here

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java [hunk 2]
```diff
@@ -407,31 +423,82 @@ public String getTablePath() {
...
 
-          totalBytes += addFile.getSize();
-          partitionedFiles.add(partitionedFile);
+    // When a limit is pushed, request stats so we can read numRecords per file.
+    // Without a limit, use the default path (stats only read if there is a data skipping filter).
+    final boolean requestStats = pushedLimit.isPresent();
```
Comment (raveeram-db) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: this section can be refactored to: `kernelScan.getScanFiles(tableEngine, pushedLimit.isPresent())`

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java [hunk 3]
```diff
@@ -407,31 +423,82 @@ public String getTablePath() {
...
+    final CloseableIterator<FilteredColumnarBatch> scanFileBatches =
+        requestStats
+            ? kernelScan.getScanFiles(tableEngine, true /* includeStats */)
+            : kernelScan.getScanFiles(tableEngine);
```
Comment (raveeram-db) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: one decision to make here is whether the limit should be pushed down to the Kernel as well? cc: @nicklan @scottsand-db

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java (hunk 1 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java:435]: Will do

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java (hunk 2 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: Thanks, good catch! Will do.

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java (hunk 1 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java:435]: Done

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java (hunk 2 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: Done.

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java (hunk 3 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: Good point. I think doing this on the connector-side is the right place for this for now.

`getScanFiles` returns a lazy iterator. We already break out early once enough rows accumulate, so no wasted work. Moving this into Kernel means adding `ScanBuilder.withLimit(int)` to the public API, plus the same DV cardinality / missing stats logic we already have here but inside `ScanImpl` (please correct me if I am wrong). While the net effect is the same, it ends up becoming a public API commitment.

If multiple connectors end up duplicating this logic, we should centralize it. Also, I *think* if we do decide to push it into Kernel later, that's a purely additive API change that won't require reworking this PR. So technically, we do not need to make a decision now. 

WDYT?

File: kernel/kernel-api/src/main/java/io/delta/kernel/Scan.java [hunk 4]
```diff
@@ -106,6 +106,28 @@ public interface Scan {
...
+   * <p>When {@code includeStats=false}, this method behaves identically to {@link
+   * #getScanFiles(Engine)}.
+   *
+   * @param engine {@link Engine} instance to use in Delta Kernel.
+   * @param includeStats whether to read and include the JSON file statistics in the returned rows.
+   * @return iterator of {@link FilteredColumnarBatch}s where each selected row corresponds to one
+   *     scan file.
+   * @since 3.3.0
+   */
+  default CloseableIterator<FilteredColumnarBatch> getScanFiles(
```
Comment (murali-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/Scan.java:131]: This adds a new `default` method to the public `@Evolving` `Scan` interface — a public API commitment. A few concerns that need kernel owners' sign-off:

1. **Boolean parameter is fragile on a public API.** If future needs arise (e.g., include column-level stats, include partition stats), the API can't be extended without another overload. A pattern like `getScanFiles(Engine, ScanFileOptions)` or an enum would be more future-proof. Changing this later is a breaking change.

2. **Default impl silently degrades.** The default returns `getScanFiles(engine)` which silently ignores `includeStats=true`. The caller (`SparkScan`) treats missing stats as "include file but don't count toward limit" — correct but the optimization silently degrades with no way for the caller to detect it. Consider throwing `UnsupportedOperationException` for `includeStats=true` in the default (as was done in #6330), or at minimum documenting the degradation.

3. **`@since 3.3.0`** — should this be `4.3.0`? (Same issue flagged and fixed in #6330.)

This needs explicit sign-off from kernel API owners before merging.

Comment (rapoth) [reply kernel/kernel-api/src/main/java/io/delta/kernel/Scan.java:131]: Rebased this PR onto `kernel-scan-getstats-api` (#6330), which already addresses concerns 2 and 3:

- **`@since` version**: Fixed to `4.3.0` (commit 79461a8).
- **Default impl**: Now throws `UnsupportedOperationException("This Scan implementation does not support includeStats=true.")` instead of silently degrading. `ScanInterfaceSuite` verifies this behavior (commit 0c44e33).

On **boolean parameter fragility**: valid concern. The `@Evolving` annotation on `Scan` gives us room to evolve this. A single boolean is the simplest thing for the one current use case (`includeStats`). If a second option arises (e.g., partition stats, max files), we can introduce a `ScanFileOptions` builder at that point without a breaking change. Premature abstraction risks guessing wrong about what future options look like.

Agreed this needs kernel API owner sign-off, which is tracked on #6330.

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java [hunk 5]
```diff
@@ -407,36 +427,99 @@ public String getTablePath() {
...
+    // Without a limit, use the default path (stats only read if there is a data skipping filter).
+    final CloseableIterator<FilteredColumnarBatch> scanFileBatches =
+        kernelScan.getScanFiles(tableEngine, pushedLimit.isPresent());
+
+    long accumulatedRecords = 0;
+    boolean limitReached = false;
+
+    try {
+      outer:
+      while (scanFileBatches.hasNext()) {
```
Comment (raveeram-db) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: Could you please add a couple of tests similar to https://github.com/delta-io/delta/blob/e8cffee7b45701910d0a6bd7a9edb26e6fcaae8c/spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkMicroBatchStreamTest.java#L228

that test for parity between DSv1 and DSv2? Would be super helpful in making sure we're not missing any edge cases here especially as we account for DV cardinality and such

**Comment by raveeram-db** [comment]
Overall, looks good, just one comment to make sure we add integration tests that validate parity between DSv1 and DSv2

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java (hunk 5 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: Thanks, updated.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+    for (int i = 0; i < v1Rows.size(); i++) {
+      assertEquals(
+          v1Rows.get(i).toString(),
+          v2Rows.get(i).toString(),
+          str("Row %d differs between V1 and V2", i));
+    }
+  }
+
+  @Test
+  public void testParity_basicMultipleFiles(@TempDir File tempDir) {
```
Comment (raveeram-db) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java]: a lot of these tests can perhaps be consolidated: 
e.g. each of the correctness and robustness tests can also run the same query with V1 and validate that the output is the same - and remove the parity tests

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScanBuilder.java [hunk 7]
```diff
@@ -165,19 +170,44 @@ public Filter[] pushedFilters() {
...
+            snapshotManager,
+            initialSnapshot,
+            dataSchema,
+            partitionSchema,
+            requiredDataSchema,
+            pushedKernelPredicates,
+            dataFilters,
+            kernelScanBuilder.build(),
+            catalogStats,
+            options);
```
Comment (huan233usc) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScanBuilder.java]: Can we directly make pushedLimit to the constructor of `SparkScan`. (Actually having a pojo called SparkScanContext for all these args might be better to prevent too many arg case)

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java [hunk 8]
```diff
@@ -407,36 +427,99 @@ public String getTablePath() {
...
+    // When a limit is pushed, request stats so we can read numRecords per file.
+    // Without a limit, use the default path (stats only read if there is a data skipping filter).
+    final CloseableIterator<FilteredColumnarBatch> scanFileBatches =
+        kernelScan.getScanFiles(tableEngine, pushedLimit.isPresent());
+
+    long accumulatedRecords = 0;
+    boolean limitReached = false;
+
+    try {
+      outer:
```
Comment (huan233usc) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: It seems very nested, and not human readable to me. Try to split it into small helpers like


```suggestion
private void planScanFiles() {
  final Engine tableEngine = DefaultEngine.create(hadoopConf);
  final String tablePath = getTablePath();

  if (pushedLimit.isPresent() && pushedLimit.getAsInt() == 0) {
    estimatedSizeInBytes = 0;
    return;
  }

  final CloseableIterator<FilteredColumnarBatch> scanFileBatches =
      kernelScan.getScanFiles(tableEngine, pushedLimit.isPresent());

  long accumulatedRecords = 0;
  try {
    while (!isLimitReached(accumulatedRecords) && scanFileBatches.hasNext()) {
      accumulatedRecords =
          processScanBatch(scanFileBatches.next(), tablePath, accumulatedRecords);
    }
  } finally {
    closeQuietly(scanFileBatches);
  }

  logLimitPushdownResult(accumulatedRecords);
  estimatedSizeInBytes = computeEstimatedSizeWithColumnProjection(totalBytes);
}

private long processScanBatch(
    FilteredColumnarBatch batch, String tablePath, long accumulatedRecords) {
  try (CloseableIterator<Row> rows = batch.getRows()) {
    while (!isLimitReached(accumulatedRecords) && rows.hasNext()) {
      final AddFile addFile = new AddFile(rows.next().getStruct(0));
      totalBytes += addFile.getSize();
      partitionedFiles.add(
          PartitionUtils.buildPartitionedFile(addFile, partitionSchema, tablePath, zoneId));
      accumulatedRecords += getLogicalRowCount(addFile);
    }
  } catch (IOException e) {
    throw new RuntimeException(e);
  }
  return accumulatedRecords;
}

/**
 * Returns the logical (surviving) row count for a file. Subtracts DV cardinality from
 * physical numRecords. Returns 0 when stats are missing (conservative: file is included
 * but not counted toward the limit).
 */
private long getLogicalRowCount(AddFile addFile) {
  Optional<Long> numRecords = addFile.getNumRecords();
  if (!numRecords.isPresent()) return 0;

  long dvCardinality = addFile.getDeletionVector()
      .map(DeletionVectorDescriptor::getCardinality)
      .orElse(0L);
  return Math.max(0, numRecords.get() - dvCardinality);
}

private boolean isLimitReached(long accumulatedRecords) {
  return pushedLimit.isPresent() && accumulatedRecords >= pushedLimit.getAsInt();
}
```

File: spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+ * reading through the DSV2 connector, and that the limit pushdown optimization reduces planned
+ * files when possible.
+ *
+ * <p>Note: LIMIT without ORDER BY is non-deterministic per SQL standard -- the specific rows
+ * returned may vary across runs. Tests therefore check row counts, not specific row values.
+ */
+public class V2LimitPushdownTest extends V2TestBase {
+
+  // ==========================================================================
+  // Priority 1: Correctness tests
```
Comment (huan233usc) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java]: Even if we don't implement the supports pushdown limit, the test will still pass. Should we consider adding some detailed inspection utils on the plan to make sure limit is really pushed down and we still return correct result?
like --
private void assertLimitPushed(String query, int expectedLimit) {
  String plan = spark.sql(query).queryExecution().executedPlan().toString();
  assertThat(plan).contains("PushedLimit: " + expectedLimit);
}

File: spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java (hunk 6 above)
Comment (rapoth) [reply spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java]: Done. Each correctness and robustness test now calls `assertV1V2Parity` (or `assertV1V2ParityWithFilter`) inline, so every test validates both count correctness and V1/V2 result parity in a single pass. The separate parity test section has been removed. See commit 92f53f3.

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScanBuilder.java (hunk 7 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScanBuilder.java]: Done. `pushedLimit` is now a `final` constructor parameter on `SparkScan`, and `build()` passes it directly. The `setPushedLimit` setter has been removed. I held off on the `SparkScanContext` POJO for now since there's only one new parameter, but happy to revisit if the arg list grows further. See commit 92f53f3.

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java (hunk 8 above)
Comment (rapoth) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/read/SparkScan.java]: Thanks for the detailed suggestion! Applied. See commit 92f53f3.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java (hunk 9 above)
Comment (rapoth) [reply spark/v2/src/test/java/io/delta/spark/internal/v2/V2LimitPushdownTest.java]: Good call. Added `assertLimitPushed(tablePath, expectedLimit)` which inspects the physical plan string for `PushedLimit: <N>`, and `assertLimitNotPushed(tablePath, filter, limit)` which verifies limit is NOT pushed when data filters produce post-scan residuals. Every applicable test now asserts on the plan, not just the result count. See commit 92f53f3.

Also added two defensive guards in a follow-up commit (a021298):
- `pushLimit` now rejects negative values with `IllegalArgumentException`
- `build()` defensively clears `pushedLimit` when post-scan data filters exist, so the connector does not silently depend on Spark's optimizer invariant
